Cabal/Distribution/Lexer.x

```diff
@@ -0,0 +1,244 @@
+{
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Distribution.Parsec.Lexer
+-- License     :  BSD3
+--
+-- Maintainer  :  cabal-devel@haskell.org
+-- Portability :  portable
+--
+-- Lexer for the cabal files.
+{-# LANGUAGE BangPatterns #-}
+module Distribution.Lexer
+  (ltest, lexToken, Token(..), LToken(..)
+  ,bol_section, in_section, in_field_layout, in_field_braces
+  ,mkLexState) where
+
+import Distribution.Parsec.LexerMonad
+import Control.Monad
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Char8 as B.Char8
+import Data.Word (Word8)
+import qualified Data.Char
+import Data.Char (chr, ord)
+import Data.List (stripPrefix)
+--import Distribution.Simple.Utils (fromUTF8) --don't decode, keep as ByteString
+
+-- testing only:
+import Debug.Trace
+import Control.Exception (assert)
+import qualified Data.Vector as V
+import qualified Data.Text   as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Text.Encoding.Error as T
+}
+
+-- Various character classes
+
+$space           = \          -- single space char
+$digit           = 0-9        -- digits
+$alpha           = [a-z A-Z]  -- alphabetic characters
+$symbol          = [\= \< \> \+ \* \- \& \| \! \$ \% \^ \@ \# \? \/ \\ \~]
+$ctlchar         = [\x0-\x1f \x7f]
+$printable       = \x0-\x10ffff # $ctlchar   -- so no \n \r
+$spacetab        = [$space \t]
+$bom             = \xfeff
+$nbsp            = \xa0
+$nbspspacetab    = [$nbsp $space \t]
+
+$paren           = [ \( \) \[ \] ]
+$field_layout    = [$printable \t]
+$field_layout'   = [$printable] # [$space]
+$field_braces    = [$printable \t] # [\{ \}]
+$field_braces'   = [$printable] # [\{ \} $space]
+$comment         = [$printable \t]
+$namecore        = [$alpha]
+$nameextra       = [$namecore $digit \- \_ \. \']
+$instr           = [$printable $space] # [\"]
+$instresc        = $printable
+
+@nl          = \n | \r\n | \r
+@name        = $nameextra* $namecore $nameextra*
+@string      = \" ( $instr | \\ $instresc )* \"
+@numlike     = $digit [$digit \.]*
+@oplike      = [ \, \. \= \< \> \+ \* \- \& \| \! \$ \% \^ \@ \# \? \/ \\ \~ ]+
+
+tokens :-
+
+<0> {
```

Comment:
FWIW, I'm pretty sure we have a check to make sure there is no leading BOM in Cabal files. So if we record this in the result of parse we can do the check purely rather than by re-reading the Cabal file and looking for a BOM. See `checkCabalFileBOM` in `Cabal/Distribution/PackageDescription/Check.hs`